pandas/io/html.py

```diff
@@ -342,44 +349,155 @@ def _build_doc(self):
+                    self._extract_tfoot(table_html)[0])
+        else:
+            # Otherwise we need to split the body into header/body/foot.
+            body_rows = self._extract_tr(table_html)
+            if body_rows == []:
+                # empty table, just return nothing
+                return [], [], []
+            # splitting criterion: if all tags within a row are th, it's part
+            # of the header/footer
+            while all(self._equals_tag(t, 'th') for t in
+                      self._extract_td(body_rows[0])):
+                # this row should be a header row, move it from body to header
+                header_rows.append(body_rows.pop(0))
+            while all(self._equals_tag(t, 'th') for t in
+                      self._extract_td(body_rows[-1])):
+                # this row should be a footer row, move it from body to footer
+                footer_rows.insert(0, body_rows.pop())
+
+        header = self._expand_colspan_rowspan(header_rows, fill_rowspan=False)
+        body = self._expand_colspan_rowspan(body_rows, fill_rowspan=True)
+        footer = self._expand_colspan_rowspan(footer_rows, fill_rowspan=False)
+
+        # The below line is lifted from _parse_raw_tfoot. Not sure what
+        # it does.
+        footer = np.atleast_1d(np.array(footer).squeeze(
+        )) if footer and len(footer) == 1 else footer
+        return header, body, footer
+
+    def _expand_colspan_rowspan(self, rows, fill_rowspan=True):
+        """Given a list of rows, return a list of rows that properly handle
+           colspan/rowspan
+
+        Discussion on behavior of fill_rowspan in #17073
+
+        Parameters
+        ----------
+        rows : list of rows, each of which is a list of elements in that row
+
+        fill_rowspan : boolean
+            Should a rowspan fill every item in the rowspan (True) or only the
+            bottommost element (False)? Default is True.
+
+        Returns
+        -------
+        res : list of rows, each of which is a list of elements in that row,
+            respecting colspan/rowspan
+        """
+
         res = []
-        if thead:
-            trs = self._parse_tr(thead[0])
-            for tr in trs:
-                cols = lmap(self._text_getter, self._parse_td(tr))
-                if any([col != '' for col in cols]):
-                    res.append(cols)
+        saved_span = []
+        for row in rows:
+            extracted_row = self._extract_td(row)
+            cols_text = [_remove_whitespace(
+                self._text_getter(col)) for col in extracted_row]
+            col_colspans = [int(col.get('colspan', 1))
+                            for col in extracted_row]
+            col_rowspans = [int(col.get('rowspan', 1))
+                            for col in extracted_row]
+            # expand cols using col_colspans
+            # maybe this can be done with a list comprehension, dunno
+            cols = list(zip(
```

Comment:
I think this is equivalent to

```python
list(chain.from_iterable([text] * nc for text, nc in zip(cols_text, col_colspans)))
```

which I find a bit more readable. Can you confirm if that works? (import `chain` from itertools).

Reply:
Hmm, reading a bit further, maybe not? But I think that's close. Could you post a sample of `col_colspans`, `col_rowspans`, and `cols_texts`? Are they all just flat lists of ints / strs?